math/number_of_increasing_sequences_between_two_sequences/sol/correct.cpp

```diff
@@ -0,0 +1,1202 @@
+            for (int i = 0; i < p; i++) {
+                auto l = a[i + offset];
+                auto r = a[i + offset + p] * now;
+                a[i + offset] = l + r;
+                a[i + offset + p] = l - r;
+            }
+            now *= sum_e[bsf(~(unsigned int)(s))];
+        }
+    }
+}
+
+template <class mint, internal::is_static_modint_t<mint>* = nullptr>
+void butterfly_inv(std::vector<mint>& a) {
+    static constexpr int g = internal::primitive_root<mint::mod()>;
+    int n = int(a.size());
+    int h = internal::ceil_pow2(n);
+
+    static bool first = true;
+    static mint sum_ie[30];  // sum_ie[i] = es[0] * ... * es[i - 1] * ies[i]
+    if (first) {
+        first = false;
+        mint es[30], ies[30];  // es[i]^(2^(2+i)) == 1
+        int cnt2 = bsf(mint::mod() - 1);
+        mint e = mint(g).pow((mint::mod() - 1) >> cnt2), ie = e.inv();
+        for (int i = cnt2; i >= 2; i--) {
+            es[i - 2] = e;
+            ies[i - 2] = ie;
+            e *= e;
+            ie *= ie;
+        }
+        mint now = 1;
+        for (int i = 0; i <= cnt2 - 2; i++) {
+            sum_ie[i] = ies[i] * now;
+            now *= es[i];
+        }
+    }
+
+    for (int ph = h; ph >= 1; ph--) {
+        int w = 1 << (ph - 1), p = 1 << (h - ph);
+        mint inow = 1;
+        for (int s = 0; s < w; s++) {
+            int offset = s << (h - ph + 1);
+            for (int i = 0; i < p; i++) {
+                auto l = a[i + offset];
+                auto r = a[i + offset + p];
+                a[i + offset] = l + r;
+                a[i + offset + p] =
+                    (unsigned long long)(mint::mod() + l.val() - r.val()) *
+                    inow.val();
+            }
+            inow *= sum_ie[bsf(~(unsigned int)(s))];
+        }
+    }
+}
+
+template <class mint, internal::is_static_modint_t<mint>* = nullptr>
+std::vector<mint> convolution_naive(const std::vector<mint>& a, const std::vector<mint>& b) {
+    int n = int(a.size()), m = int(b.size());
+    std::vector<mint> ans(n + m - 1);
+    if (n < m) {
+        for (int j = 0; j < m; j++) {
+            for (int i = 0; i < n; i++) {
+                ans[i + j] += a[i] * b[j];
+            }
+        }
+    } else {
+        for (int i = 0; i < n; i++) {
+            for (int j = 0; j < m; j++) {
+                ans[i + j] += a[i] * b[j];
+            }
+        }
+    }
+    return ans;
+}
+
+template <class mint, internal::is_static_modint_t<mint>* = nullptr>
+std::vector<mint> convolution_fft(std::vector<mint> a, std::vector<mint> b) {
+    int n = int(a.size()), m = int(b.size());
+    int z = 1 << internal::ceil_pow2(n + m - 1);
+    a.resize(z);
+    internal::butterfly(a);
+    b.resize(z);
+    internal::butterfly(b);
+    for (int i = 0; i < z; i++) {
+        a[i] *= b[i];
+    }
+    internal::butterfly_inv(a);
+    a.resize(n + m - 1);
+    mint iz = mint(z).inv();
+    for (int i = 0; i < n + m - 1; i++) a[i] *= iz;
+    return std::move(a);
```

Comment:
```suggestion
    return a;
```

古いバージョンの AtCoder Library を使っていることが元の原因の 1 つです。この問題では出力の正しさは別でチェックするとして、今後はなるべく新しいバージョンだと助かります。

Reply:
分かりました。ありがとうございます。